### PR TITLE
feat(payment): INT-7169 BlueSnapDirect: Add payment method component

### PR DESCRIPTION
**Pull request by mauricio-sg**
## What? [INT-7169](https://bigcommercecloud.atlassian.net/browse/INT-7169)
Add payment method component for BlueSnap direct ACH/ECP.

**NOTE:** This PR was created as a result of having closed this other 👉 #1208

## Why?
It's required to collect the shopper's payment data.

## Testing / Proof
Screenshot 2023-03-01 at 3 45 51 PM

## Depends on
👉 https://github.com/bigcommerce/checkout-sdk-js/pull/1918

@bigcommerce/checkout

[INT-7169]: https://bigcommercecloud.atlassian.net/browse/INT-7169?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import { FormFieldItem } from '@bigcommerce/checkout-sdk';
+import classNames from 'classnames';
 import { isDate, noop } from 'lodash';
 import React, { FunctionComponent, memo, useCallback } from 'react';
 import ReactDatePicker from 'react-datepicker';
 
-import { useLocale } from '@bigcommerce/checkout/locale';
+// import { useLocale } from '@bigcommerce/checkout/locale';
```
Comment (vmparra) [inline packages/ui/src/form/DynamicFormField/DynamicInput.tsx:7]: Why is this line commented?

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx [hunk 2]
```diff
@@ -29,11 +32,14 @@ const DynamicInput: FunctionComponent<DynamicInputProps> = ({
     onChange = noop,
     options,
     placeholder,
+    useFloatingLabel,
     value,
     ...rest
 }) => {
-    const { date } = useLocale();
-    const { inputFormat } = date || { inputFormat: '' };
+    // FIXME: useLocale must be used within a LocaleContextProvider
```
Comment (vmparra) [inline packages/ui/src/form/DynamicFormField/DynamicInput.tsx:39]: Are we planning to make the fix as part of this PR?

**Review by vmparra — COMMENTED** [review]
Hi @mauricio-sg, it looks great so far. Just left a couple of questions.

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx (hunk 2 above)
Comment (mauricio-sg) [reply packages/ui/src/form/DynamicFormField/DynamicInput.tsx:39]: `locale` is not working right now, @bc-peng told me that there are several tickets pending to make it work.

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx (hunk 1 above)
Comment (mauricio-sg) [reply packages/ui/src/form/DynamicFormField/DynamicInput.tsx:7]: Please see the comment below.

**Review by vmparra — APPROVED** [review]
LGTM 👍🏻

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx [hunk 3]
```diff
@@ -29,11 +32,14 @@ const DynamicInput: FunctionComponent<DynamicInputProps> = ({
...
+    // const { date } = useLocale();
```
Comment (animesh1987) [inline packages/ui/src/form/DynamicFormField/DynamicInput.tsx:40]: Is this needed ?

File: packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx [hunk 4]
```diff
@@ -276,5 +276,6 @@ export default toResolvableComponent<PaymentMethodProps, PaymentMethodResolveId>
             id: 'hosted-credit-card',
             // type: 'PAYMENT_TYPE_API',
```
Comment (animesh1987) [inline packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx:277]: Can I ask why is this bit commented out as well? Also these are supposed to be helper components. What I mean by that is that `BluesnapComponent` package will import and will be composed using this.

File: packages/bluesnap-direct-integration/src/validation-schemas/getEcpValidationSchema.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    return object({
+        accountNumber: string()
+            .required('Account Number is required')
+            .matches(/^\d+$/, 'Account Number must contain only numbers')
+            .min(8, 'Account Number must have at least 8 digits'),
+        routingNumber: string()
+            .required('Routing Number is required')
+            .matches(/^\d+$/, 'Routing Number must contain only numbers')
+            .length(9, 'Routing Number must be 9 digits long'),
+        accountType: string().required('Account Type is required'),
```
Comment (animesh1987) [inline packages/bluesnap-direct-integration/src/validation-schemas/getEcpValidationSchema.ts:14]: Do we need to look into translating these?

File: packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx (hunk 4 above)
Comment (mauricio-sg) [reply packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx:277]: > Can I ask why is this bit commented out?

**A:** It's not my change but I found the answer 👉 https://github.com/bigcommerce/checkout-js/pull/1146#discussion_r1065220457

> Also these are supposed to be helper components.

**Q:** Do you mean that I am implementing it wrong? Could you give me an example of how to use it correctly? So, what's the purpose of being a resolvable component? I thought it was intended to work the same way as new payment strategies.

File: packages/bluesnap-direct-integration/src/validation-schemas/getEcpValidationSchema.ts (hunk 5 above)
Comment (mauricio-sg) [reply packages/bluesnap-direct-integration/src/validation-schemas/getEcpValidationSchema.ts:14]: Please see comments above. https://github.com/bigcommerce/checkout-js/pull/1218#discussion_r1147850530 and https://github.com/bigcommerce/checkout-js/pull/1218#discussion_r1123709898

**Comment by mauricio-sg** [comment]
@animesh1987 ♻️ 🙏

File: packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx (hunk 4 above)
Comment (bc-peng) [reply packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx:277]: Hi @mauricio-sg , Animesh and I talked about it, and it came to our attention that you edited this package.

Ideally, with the helper CC component(https://github.com/bigcommerce/checkout-js/tree/master/packages/credit-card-integration), you should be able to build your own credit card payment method component similar to the `hosted-credit-card-integration`. You don't need to touch this file.

Furthermore, as a new payment package, we highly suggest you to have an e2e test to ensure that no future modifications could break it.

Resources:
- e2e test code sample: https://github.com/bigcommerce/checkout-js/blob/master/packages/barclay-integration/e2e/barclay.spec.ts
- e2e testing Doc: https://bc-eng-docs.herokuapp.com/domains/checkout/payment-integrations-contributing-guide-checkout.html#e2e-test

**Comment by animesh1987** [comment]
Hey @mauricio-sg can you please address this comment from Peng
> Furthermore, as a new payment package, we highly suggest you to have an e2e test to ensure that no future modifications could break it.

Resources:

e2e test code sample: https://github.com/bigcommerce/checkout-js/blob/master/packages/barclay-integration/e2e/barclay.spec.ts
e2e testing Doc: https://bc-eng-docs.herokuapp.com/domains/checkout/payment-integrations-contributing-guide-checkout.html#e2e-test

**Comment by mauricio-sg** [comment]
Comments have been addressed, but the ci/circleci: e2e will not pass until the dependent PR is merged. ☝️😬

**Comment by david-intersys** [comment]
> > Comments have been addressed, but the ci/circleci: e2e will not pass until the dependent PR is merged.
> 
> Is dependant PR the one that's in draft for SDK? Is that is the case we should focus on merging that first. I remember it might be blocked by a bug, we should be able to address the bug in next couple of days.

@animesh1987 Any news about this?

**Comment by animesh1987** [comment]
Hey @david-intersys we merged the PR for SDK fix today https://github.com/bigcommerce/checkout-sdk-js/pull/1946. If @mauricio-sg rebases checkout js to latest, the change should be there. Apologies for the delays.

**Comment by BC-BMedina** [comment]
💚 Tested!
QE Details:
**ACH/ECP** alternative payment method with **Bluesnap Direct** has been validated under the following scenarios:

- Approved orders
- Declined orders
- Format validation
- USD Currency exclusivity
- Place orders with credit card
- Place Orders with 3DS credit card

They are working as expected, here you will find the evidence of the tests:
[INT-7169 [AMETEK MVP] Display ACH as available payment method in the checkout.pdf](https://drive.google.com/file/d/1ozOY4caaomjUYUr9_jCWsa8HCPZMv4GK/view?usp=sharing)

**Comment by mauricio-sg** [comment]
@animesh1987 @bc-peng This is ready for review now ♻️ 🙏

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx (hunk 1 above)
Comment (bc-peng) [reply packages/ui/src/form/DynamicFormField/DynamicInput.tsx:7]: Could you please update this line?

Comment (mauricio-sg) [reply packages/ui/src/form/DynamicFormField/DynamicInput.tsx:7]: @bc-peng what do you mean? Is `useLocale` working now?

Comment (bc-peng) [reply packages/ui/src/form/DynamicFormField/DynamicInput.tsx:7]: Yes, we have finalised the `locale` package.
The implementation is at https://github.com/bigcommerce/checkout-js/pull/1245.

File: packages/bluesnap-direct-integration/e2e/bluesnapdirect.spec.ts
Comment (bc-peng) [inline packages/bluesnap-direct-integration/e2e/bluesnapdirect.spec.ts:1]: 👍

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx
Comment (bc-peng) [inline packages/ui/src/form/DynamicFormField/DynamicInput.tsx:1]: Could you update the comments of this file based on the lastest checkout-js?

File: packages/core/src/app/payment/PaymentForm.tsx [hunk 6]
```diff
@@ -217,6 +217,8 @@ const PaymentMethodListFieldset: FunctionComponent<PaymentMethodListFieldsetProp
                 paymentProviderRadio: getUniquePaymentMethodId(method.id, method.gateway),
                 shouldCreateAccount: true,
                 shouldSaveInstrument: false,
+                accountNumber: '',
+                routingNumber: '',
```
Comment (bc-peng) [inline packages/core/src/app/payment/PaymentForm.tsx:221]: Could you please link me the related type/interface changes?

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx (hunk 3 above)
Comment (animesh1987) [reply packages/ui/src/form/DynamicFormField/DynamicInput.tsx:40]: This should be fixed now, you might need to rebase on the latest master.

File: packages/core/src/app/payment/PaymentForm.tsx (hunk 6 above)
Comment (mauricio-sg) [reply packages/core/src/app/payment/PaymentForm.tsx:221]: https://github.com/bigcommerce/checkout-js/blob/a55aa556b13c75c527e8394374f24947d32e0f3c/packages/payment-integration-api/src/PaymentFormValues.ts#L2

File: packages/ui/src/form/DynamicFormField/DynamicInput.tsx
Comment (mauricio-sg) [reply packages/ui/src/form/DynamicFormField/DynamicInput.tsx:1]: done!

**Comment by mauricio-sg** [comment]
Rebased and updated @bc-peng ♻️ 🙏

**Review by bc-peng — APPROVED** [review]
LGTM. Nice work on the new package and e2e test.

File: packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx (hunk 4 above)
Comment (animesh1987) [reply packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx:277]: Going back to this if this is still commented, can we remove this?

Comment (animesh1987) [reply packages/hosted-credit-card-integration/src/HostedCreditCardPaymentMethod.tsx:277]: @mauricio-sg

**Pull request merged**